### ISSUE TITLE
Branch: main

### ISSUE THREAD

**Issue by r2d2Proton** [ocornut/imgui#6141]
Any chance you can switch to "main" branch. "master" is viewed more as a derogatory term.

**Comment by ocornut** [comment]
I am waiting for git+github advertised redirection feature. Without it every PR would immediatly be orphaned/closed and all existing pointers to this repo would be broken.

The new imgui_test_engine already use main.

**Comment by GamingMinds-DanielC** [comment]
I just took a look at what else you "contributed" to different repositories, I couldn't find anything except for creating issues about master-branches. In one of them you straight out accused the owner of the repository of being derogatory, that's quite rude. I also never encountered anyone before that saw "master" as a problematic term. There was the debate about "master/slave", but the problem (if you want to call it one) there was "slave", not "master".

As someone else already told you, "master" has no inherent derogatory meaning. Of course every repository owner can handle this however they prefer, but this is a non-problem and from my perspective any effort put into this is wasted time. There are more than enough useful things that could be done instead.

**Comment by ocornut** [comment]
I'd rather not make this place a thread for discussions that have happened a million times elsewhere on the internet, but I am generally in the favor that social justice requires from us a readjustment of habits in many areas and that includes vocabulary. If we are not a victim of a given problem our voice doesn't matter much when it comes to decide if that thing is important. There are hundreds millions people who are direct victims of the long-running effect of slavery, so personally I think it matters. At the end of the day, even if it doesn't seem important in the grand scheme of them, by agreeing or disagreeing to this desire we are taking a stance for or against that element of justice. There are enough real devils in this world that we don't need to be playing devil's advocate.

As repository owner I would all be in favor of renaming to "main" if it wasn't for the technical issues it currently opens. 

It has been a big topic in git, which AFAIK so far offers the choice for new repo but couldn't offers a real redirection system.

**Comment by ocornut** [comment]
interestingly accordingly to https://github.com/github/renaming, PRs are retargeted, so it may be less problematic I thought it would be. It would mostly leave old pointers dangling but if we push an explicit commit on top of old branch name that would increase visibility.

**Comment by PathogenDavid** [comment]
I think if you re-add a `master` branch after the rename in order to add that explicit commit you'll end up breaking some of GitHub's renaming features (such as redirecting old `master` links to `main`.)

For people finding out locally, I like to think most people will notice `origin/master` has disappeared and figure out what's up. If not they should eventually find out from release notes.

**Comment by ocornut** [comment]
Sorry I meant pushing a last commit _BEFORE_ the rename, so that's the commit that appears if people update to latest to master.

**Comment by PathogenDavid** [comment]
The problem is they wouldn't see that since `origin/master` no longer exists and as such their local reference for it will no longer be updated. It'll be stuck on wherever it was last time they fetched the repo.

So unless you intend to commit that message and then commit nothing else for a while, I don't think a lot of the target audience would even see it.

You do get warning from Git though:

```
C:\Scratch\BranchRenameTest\Two>git pull origin
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 1 (delta 0), pack-reused 0
Unpacking objects: 100% (1/1), 867 bytes | 216.00 KiB/s, done.
From https://github.com/PathogenPlayground/BranchRenameTest
 * [new branch]      main       -> origin/main
Your configuration specifies to merge with the ref 'refs/heads/master'
from the remote, but no such ref was fetched.
```

which is slightly drowned out by the other info but if they ever pull with no remote changes it's more obvious:

```
C:\Scratch\BranchRenameTest\Two>git pull origin
Your configuration specifies to merge with the ref 'refs/heads/master'
from the remote, but no such ref was fetched.
```

They might see it anyway (IE: if they're using a GUI for Git and notice commits on other branches easily) but at that point you're guessing what people's workflows are gonna be.

------------

On the web interface side of things it's pretty obvious though. After the rename this message appears and will continue to reappear until it's dismissed:

![Screenshot of GitHub's default branch renamed notice including commands for migrating from the old branch name.](https://user-images.githubusercontent.com/278957/217797495-66c05503-2120-4217-85ac-7a3088059de3.png)

IMO overall Dear ImGui receives such frequent development that anyone even slightly paying attention is going to notice something eventually.

**Comment by ocornut** [comment]
> The problem is they wouldn't see that since origin/master no longer exists and as such their local reference for it will no longer be updated. It'll be stuck on wherever it was last time they fetched the repo.

Right, my bad.

> IMO overall Dear ImGui receives such frequent development that anyone even slightly paying attention is going to notice something eventually.

Tend to agree.

What happened with Emscripten is:
- they didn't delete old branch but added their own message https://github.com/emscripten-core/emsdk/commit/05328c455991a50550933a8eda9cb4952af1c93d + https://github.com/emscripten-core/emsdk/commit/7e3c0461444bc2b3ad33b32fe89867d626e0f9f5
- so I didn't get the git-side error but somehow didn't see their message either.

Going to try with https://github.com/ocornut/imgui_club now.

**Comment by ocornut** [comment]
Result on imgui_club
- Out of 4 pull requests, 3 were retargeted, 1 was closed. The closed one is one for which the initial branch was deleted already (this normally keeps PR opens and GitHub still host the diff/patch).
- I consider this a GitHub bug that we should report. Will investigate what is the best reporting channel.

Before:
![image](https://user-images.githubusercontent.com/8225057/217805891-743c804f-49fc-4772-89e3-2b992cd2bf30.png)

After:
![image](https://user-images.githubusercontent.com/8225057/217806041-69de18d1-7d69-4263-a615-f94fc94501fb.png)

Closed PR:
https://github.com/ocornut/imgui_club/pull/2

The 3 out of 4 was reported in the UI before renaming.

The official guide says:
**"Renaming a branch will [...] Re-target any open pull **requests"**
Which is technically incorrect.

If I start the rename operation on `imgui` it says: "Will update 105 pull requests targeting this branch across 91 repositories. " whereas we have 140 open pull requests. So renaming would erroneously close 35 pull requests which I still consider useful and valid.

I don't know how to meaningfully report that bug to GitHub without it being lost in the noise, but it would seem practically important that GitHub fixes it.

(cc: @elmindreda)

**Comment by ocornut** [comment]
FYI i have reported the aforementioned bug to GitHub back then.
Some extracts,

----

**Reported:**
"
According to:
https://github.com/github/renaming

"Renaming a branch will [...] Re-target any open pull requests"

This is however not correct!

I tried on one of my smaller repo:
https://github.com/ocornut/imgui/issues/6141#issuecomment-1424071102
And noticed that out of 4 PR, only 3 were retargeted, 1 was closed.
1 PR had its source branch deleted already. However this normally does NOT close the PR, and github preserve the diff/patch and still allow merging of PR. Which is good. But this is lost when we attempt to rename.

- This is preventing me from renaming master to main in https://github.com/ocornut/imgui : approximately 10 valid PR would be erroneously closed.
- I consider this a bug in GitHub which could be fixed.
- It seems particularly important to fix IMHO as in the topic of switching from master to main, people have been quick to point to technical issues related to that rename. This specific issue in GitHub makes renaming a branch a more destructive operation that it should be, therefore a less attractive operation. As much as I would like to rename my branch, I cannot afford losing 10 PR.

Reproduction Steps

- UserA opens UserA/RepoA.
- UserB forks UserB/RepoA
- UserB creates a branch, makes commits, submit PR to RepoA
- UserB delete their branch or close their account.
- RepoA STILL has the PR open.
- RepoA renames branch: PR is lost.

Fault Type

Permanent - Occurring repeatably"

----

**Followup**
"A friend of mine did some additional researchs on that and suggested that GitHub have changed the behavior to when a PR branch is deleted. It seems (but I have not confirmed it myself) that sometimes in the past, when a third-party PR branch got deleted github would keep the PR open, whereas nowadays it tend to close the PR. It may be that the older PR which already had their branch deleted are the ones closed when using rename.

Either way branch rename is currently a destructive operation."

----

**Followup**
"
> it's likely not possible to sync the PR after a target branch rename while its source branch (or repository) is deleted.

Since practically forever, github did host a patch in the issue thread when the PR branch was removed. I don't see any hard technical reasons from not keeping that patch available and the PR open.

Right now:
- It doesn't say PR are going to be closed.
- There's no way to tell which ones have been closed after the fact (other than perhaps using some github API/hooks?)
- There's no way to reopen the PR.

Renaming is currently an unnecessarily destructive operation which makes it unusable."

----

**Their last answer was**

"Thanks for your patience! :bow:

The current functionality is that PRs are automatically closed when the head repo has been deleted, and also in cases where the commit history has diverged (force pushing).

In this scenario, we cannot guarantee that historic PRs meeting the above criteria will not be permanently closed should you choose to rename the branch.

While your feedback has been noted by our engineers, its low impact and frequency means other, more pressing issues will be prioritized ahead of it. With this in mind, I can't make any promises as to if or when the behavior here might change.

We also invite you to share your thoughts in our [official GitHub community discussions repository](https://github.com/community/community/discussions).

Please note that while the Product team reads and evaluates all feedback, we cannot guarantee a response to every submission."

----

TL;DR; GitHub has a bug, can close some PR on rename, doesn't allow to reopen them even though they have a system in place to keep PR open with patches when a source branch has been removed.
